### DISCUSSION TITLE
Typescript vs register and typed controller

### DISCUSSION THREAD

**Discussion by askurihin** [react-hook-form/react-hook-form#5390] [Q&A]
Hello.
Thank you for your work on react-hook-form library!

I spent last few hours figuring out how to make types happy, here is the sample code:

https://codesandbox.io/s/red-voice-23fgl?file=/src/App.tsx

`register` method returns object with `.name` property which type is a `string`,
but typed controller expects it to be one of FieldValue keys (string literals).

Is it a bug or Im I doing something wrong?

**Comment by askurihin** [comment]
I "fixed" my issue by modifying `UseFormRegisterReturn` type to be a generic (and UseFormRegister return type), like this:

react-hook-form/dist/types/form.d.ts
```typescript
export declare type UseFormRegisterReturn<TFieldName> = {
    onChange: ChangeHandler;
    onBlur: ChangeHandler;
    ref: RefCallBack;
    name: TFieldName;
};

export declare type UseFormRegister<...> => UseFormRegisterReturn<TFieldName>;

```

Is this correct?

**Comment by bluebill1049** [comment]
You can use `FieldPath` type

`FieldPath<FormValues>`

**Reply by askurihin** [reply]
Sorry, I didn't get it. Where should I use `FieldPath` type?

**Reply by bluebill1049** [reply]
```
type FormValues = {
 test: string, 
 ...
}

name: FieldPath<FormValues>
```
